app/code/Magento/CatalogRule/Model/Indexer/IndexBuilder.php

```diff
@@ -309,9 +309,9 @@ protected function doReindexByIds($ids)
     {
         $this->cleanProductIndex($ids);
 
-        /** @var Rule[] $activeRules */
-        $activeRules = $this->getActiveRules()->getItems();
-        foreach ($activeRules as $rule) {
+        $activeRules = $this->getActiveRules();
+        while ($rule = $activeRules->fetchItem()) {
```

Comment:
Looks like using fetchItem make an improvement due to the fact that it does not call \Magento\CatalogRule\Model\ResourceModel\Rule\Collection::_afterLoad and it breaks functionality that requires data is loaded in this method